pkg/safemath/safemath.go

```diff
@@ -0,0 +1,33 @@
+package safemath
+
+import "math/bits"
+
+// This is a constant-time version of the following function:
+//
+//	func SafeOffset(x uint64, y int16) (res uint64, isOverflow bool) {
+//		res = x + uint64(y)
+//		if y < 0 {
+//			isOverflow = res >= x
+//		} else {
+//			isOverflow = res < x
+//		}
+//		return
+//	}
+//
+// This shows better results because the final bytecode
+// doesn't contain any conditional jump instructions
+// making it easier for a processor to pipeline the function.
+func SafeOffset(x uint64, y int16) (res uint64, isOverflow bool) {
+	// I'll leave proving that this is correct as an exercise for the reader :)
+	res = x + uint64(y)
+	// Why does this work?
+	// Let's proceed by cases on MSB(x).
+	// If MSB(x) == 1 then we don't have any potential overflow issues
+	// if y < 0 (i.e. MSB(y) == 1), meaning that we can consider the case y >= 0.
+	// In that case we can only wrap up by going to the begining of uint64 range making the MSB(res) = 0.
+	// This is the second disjunct of the DNF.
+	// In the same fashion the case MSB(X) == 0 is reasoned about.
+	// Finally, we boil everything down to MSBs by rotating and anding with ...000001.
+	isOverflow = bits.RotateLeft64((^x&uint64(y)&res)|(x & ^uint64(y) & ^res), 1)&0x1 != 0
```

Comment:
since `uint64(y)` is repeated several times, would you consider creating a var containing it?